straxen/plugins/afterpulse_processing.py

```diff
@@ -0,0 +1,345 @@
+        # if hitfinder_thresholds config
+        elif isinstance(self.config['hit_min_amplitude'], str):
+            self.hit_thresholds = straxen.hit_min_amplitude(
+                self.config['hit_min_amplitude'])
+        else: # int or array
+            self.hit_thresholds = self.config['hit_min_amplitude']
+            
+        
+    def compute(self, raw_records):
+        
+        # Convert everything to the records data type -- adds extra fields.
+        records = strax.raw_to_records(raw_records)
+        del raw_records
+        
+        # channel split: only need the TPC PMT channels
+        r = records[records['channel'] < self.config['n_tpc_pmts']]
+        del records
+        
+        # calculate baseline and baseline rms
+        strax.baseline(r,
+                       baseline_samples=self.config['baseline_samples'],
+                       flip=True)
+ 
+        # find all hits
+        hits = strax.find_hits(r,
+                               min_amplitude = self.hit_thresholds,
+                               min_height_over_noise = self.config['hit_min_height_over_noise'],
+                              )
+        
+        # sort hits first by record_i, then by time
+        hits_sorted = np.sort(hits, order=('record_i', 'time'))
+        
+        # find LED hits and afterpulses within the same WF
+        hits_ap = find_ap(hits_sorted,
+                          r,
+                          LED_window_left=self.config['LED_window_left'],
+                          LED_window_right=self.config['LED_window_right'],
+                          hit_left_extension=self.config['hit_left_extension'],
+                          hit_right_extension=self.config['hit_right_extension'],
+                         )
+    
+        hits_ap['area_pe'] = hits_ap['area'] * self.to_pe[hits_ap['channel']]
+        hits_ap['height_pe'] = hits_ap['height'] * self.to_pe[hits_ap['channel']]
+        
+        return hits_ap
+
+
+@export
+def find_ap(hits, records, LED_window_left, LED_window_right, hit_left_extension, hit_right_extension):
+    buffer = np.zeros(len(hits), dtype=dtype_ap)
+    if not len(hits):
+        return buffer
+    res = _find_ap(hits, records, LED_window_left, LED_window_right, hit_left_extension, hit_right_extension, buffer=buffer)
+    return res
+            
+
+@numba.jit(nopython=True, nogil=True, cache=True)
+def _find_ap(hits, records, LED_window_left, LED_window_right, hit_left_extension, hit_right_extension, buffer=None):
+    # hits need to be sorted by record_i, then time!
+    
+    offset = 0
+    
+    is_LED = False
+    t_LED = None
+    
+    prev_record_i = hits[0]['record_i']
+    record_data = records[prev_record_i]['data']
+    record_len = records[prev_record_i]['length']
+    baseline_fpart = records[prev_record_i]['baseline'] % 1
+    
+    for h_i, h in enumerate(hits):
+        
+        if h['record_i'] > prev_record_i:
+            # start of a new record
```

Comment:
Please remind me here it has been sometime since we spoke. In your case a single record is equal to a single pulse is not it? E.g. pulse_length of records is equal to length of records?

Reply:
I think yes. I'm not sure what the exact definition of a 'pulse' is in strax/straxen, but in the triggered AP data every recorded single-channel WF (=record?) is always 800 samples long and contains the physical PMT pulses from LED and afterpulses. So every record corresponds to one LED trigger.

Reply:
Sounds good. Yes so a pulse in DAQ and strax/straxen jargon is one entire waveform independently if it is fragment (normal data) or not (your case.)